eventtracking/django/models.py

```diff
@@ -0,0 +1,219 @@
+        max_length=50,
+        verbose_name='Backend name',
+        null=False,
+        blank=False,
+        help_text=(
+            'Name of the tracking backend on which this filter should be applied.'
+            '<br/>'
+            'Please note that this field is <b>case sensitive.</b>'
+        )
+    )
+
+    is_enabled = models.BooleanField(
+        default=True,
+        verbose_name='Is Enabled'
+    )
+
+    filter_type = models.CharField(
+        max_length=9,
+        choices=FILTER_TYPES,
+        verbose_name='Filter type',
+        help_text=(
+            'Allowlist: Only events matching the regular expressions in the list '
+            'will be allowed to passed through.'
+            '<br/>'
+            'Blocklist: Events matching any regular expression in the list will be '
+            'blocked.')
+    )
+
+    regular_expressions = models.TextField(
+        max_length=500,
+        verbose_name='List of regular expressions',
+        help_text=('This should be a list of regular expressions, seperated '
+                   'by a newline, for the events to be filtered.'),
+        validators=(validate_regex_list, )
+    )
+
+    class Meta:
+        verbose_name = 'Regular Expression Filter'
+        ordering = ('backend_name', 'is_enabled', '-modified')
+
+    def __str__(self):
+        return '{backend} - {type} - {is_enabled}'.format(
+            backend=self.backend_name,
+            type=self.filter_type,
+            is_enabled='Enabled' if self.is_enabled else 'Disabled'
+        )
+
+    @property
+    def compiled_expressions(self):
+        """
+        Return a list of compiled regular expressions
+        """
+        key = get_cache_key(expressions=self.regular_expressions)
+        cache_response = TieredCache.get_cached_response(key)
+
+        if cache_response.is_found:
+            logger.info('Compiled expressions found in cache for filter "%s"', self)
+            return cache_response.value
+
+        logger.info('Compiled expressions not found in cache for filter "%s"', self)
+        compiled_expressions, invalid_expressions = _clean_expressions(self.regular_expressions)
+
+        if invalid_expressions:
+            logger.error(
+                'The following invalid expressions were found in the filter: %s',
+                ', '.join(invalid_expressions)
+            )
+
+        TieredCache.set_all_tiers(key, compiled_expressions)
+        logger.info('Added compiled expressions in cache for filter "%s"', self)
+        return compiled_expressions
+
+    @classmethod
+    def get_latest_enabled_filter(cls, backend_name=None):
+        """
+        Wrapper method for _get_latest_enabled_filter. First find the
+        required filter from the cache and return it if found. Otherwise
+        get the filter from DB and cache it.
+        """
+        return cls._get_cached_filter(backend_name=backend_name)
+
+    @classmethod
+    def _get_cached_filter(cls, backend_name):
```

Comment:
I don't think that all this regex caching is necessary, it's built in to the library. See here: https://stackoverflow.com/questions/12514157/how-does-pythons-regex-pattern-caching-work

That said you do need to properly call the regex and not re-create them every time you attempt to route a message.

Reply:
The cache that `re` maintains is a single process cache (and has a limit of 512 expressions). Using this `TieredCache` make sure that cache will be shared among processes.